docs/BOOT_LOADER_SPECIFICATION.md

```diff
@@ -108,37 +142,43 @@ from the user. Only entries matching the feature set of boot loader and system
-ownership of the whole file system. Boot loaders, firmware, and other software
-implementing this specification may choose to place other files and directories
-in the same file system. For example, boot loaders that implement this
-specification might install their own boot code on the same partition; this is
-particularly common in the case of the ESP. Implementations of this specification
-must be able to operate correctly if files or directories other than `/loader/`
-are found in the top level directory. Implementations that add their own files
-or directories to the file systems should use well-named directories, to make
-name collisions between multiple users of the file system unlikely.
+Note that the three partitions described above are not supposed to be the
+exclusive territory of this specification. This specification only defines
+semantics of the `/loader/entries/` directory (along with the companion file
+`/loader/entries.srel`) and the `/EFI/Linux/` directory inside the file system,
+but it doesn't intend to define ownership of the rest of the file system. Boot
+loaders, firmware, and other software implementing this specification may
+choose to place other files and directories in the same file system. For
+example, boot loaders that implement this specification might install their own
+boot code on the same partition; this is particularly common in the case of the
+ESP. Implementations of this specification must be able to operate correctly if
+files or directories other than `/loader/entries/` and `/EFI/Linux/` are found
+in the top level directory. Implementations that add their own files or
+directories to the file systems should use well-named directories, to make name
+collisions between multiple users of the file system unlikely.
 
 ### Type #1 Boot Loader Specification Entries
 
-`$ESP/loader/` is the main directory containing the configuration for the boot
-loader.
-
-**Note:** _In all cases the `/loader/` directory should be located directly in
-the root of the file system. Specifically, the `/loader/` directory should
-**not** be located under the `/EFI/` subdirectory on the ESP._
-
-`$BOOT/loader/entries/` and `$XBOOTLDR/loader/entries/` are the directories
-containing the drop-in snippets defining boot entries, one `.conf` file for
-each boot menu item. Each OS may provide one or more such entries. The boot
-loader should enumerate both directories and provide a merged list.
-
-The file name is used for identification of the boot item but shall never be
-presented to the user in the UI. The file name may be chosen freely but should
-be unique enough to avoid clashes between OS installations. More specifically,
-it is suggested to include the `entry-token` (see
+`/loader/entries/` in `$BOOT` is the primary directory containing Type #1
+drop-in snippets defining boot entries, one `.conf` file for each boot menu
+item. Each OS may provide one or more such entries.
```

Comment:
I am not convinced that retro-updating old boot entries is a good thing. It's OK to update the current kernel, but the point of keeping the old stuff around is to keep the old stuff around, not to keep some hybrid of old and new around.

So, yes, it might make sense to invoke "kernel-install" for an existign kernel, to update things, but I am very strongly opposed to having a default codepaths of updating *all* kernels like this.

(I'd be fine with adding an option to kernel-install that *on* *request* allows rerunning the tool for all existing kernels (e.g. `kernel-install -a`), regenerating all snippets, but that should be an option for advanced admins/hackers, not part of the regular code flow)

Reply:
I think don't think the spec needs to say more here. In general, the expectation is that normal updates just add new entries, and the spec describes how to pick file names to avoid conflicts. But there may be situations where all entries for a given OS should be edited, or maybe even a situation where editing entries for other OSes makes sense. So any questions about what to update and when are left to the implementation.